### PR TITLE
fix(rustdoc): generics search

### PR DESCRIPTION
**Pull request by notriddle**
This commit adds a test case for generics, re-adds generics data
to the search index, and tweaks function indexing to use less space in JSON.

This partially reverts commit 14ca89446c076bcf484d3d05bd991a4b7985a409.

### CODE REVIEW

File: src/librustdoc/html/static/search.js [hunk 1]
```diff
@@ -106,7 +106,7 @@ function levenshtein(s1, s2) {
 window.initSearch = function(rawSearchIndex) {
     var MAX_LEV_DISTANCE = 3;
     var MAX_RESULTS = 200;
-    var GENERICS_DATA = 1;
+    var GENERICS_DATA = 2;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/search.js:109]: Why did you change this?

**Comment by GuillaumeGomez** [comment]
Overall looks good to me, thanks a lot for this fix! Just one thing: you can put multiple search index tests into one file (like it's done in [here](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc-js/doc-alias.js)). I made this comment because: can you add a test where you only look for the generic parameter please?

**Comment by GuillaumeGomez** [comment]
Also: I'll need to make a perf run because some types' increased, so we have a track of it.

File: src/librustdoc/html/static/search.js (hunk 1 above)
Comment (notriddle) [reply src/librustdoc/html/static/search.js:109]: Because field 1 has the TypeKind in it. That’s why it didn’t work.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/search.js:109]: Arf, side-effect from not enough tests... Thanks!

**Comment by notriddle** [comment]
> I made this comment because: can you add a test where you only look for the generic parameter please?

Sure. `generics` and `generics-trait` now both check with the "just the parameter" case.

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -95,7 +95,31 @@ crate struct IndexItem {
...
+    idx: Option<usize>,
     name: Option<String>,
+    generics: Option<Vec<Generic>>,
+}
+
+/// A type used for the search index.
+#[derive(Debug)]
+crate struct Generic {
+    name: String,
+    defid: Option<DefId>,
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:108]: not: `def_id` instead please. :)

File: src/librustdoc/html/render/mod.rs [hunk 3]
```diff
@@ -95,7 +95,31 @@ crate struct IndexItem {
 /// A type used for the search index.
 #[derive(Debug)]
 crate struct RenderType {
+    ty: Option<DefId>,
+    idx: Option<usize>,
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:99]: I can only find `idx` set to `None`. Am I missing something maybe? Otherwise please remove this field.

File: src/librustdoc/html/render/cache.rs [hunk 4]
```diff
@@ -251,6 +252,19 @@ fn get_index_type_name(clean_type: &clean::Type, accept_generic: bool) -> Option
...
 
+fn get_generics(clean_type: &clean::Type) -> Option<Vec<Generic>> {
+    clean_type.generics().and_then(|types| {
+        let r = types
+            .iter()
+            .filter_map(|t| {
+                get_index_type_name(t, false)
+                    .map(|name| Generic { name: name.as_str().to_ascii_lowercase() })
+            })
+            .collect::<Vec<_>>();
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/cache.rs:263]: What happens in case there is the same type used in generics multiple times:

```
fn foo<T: Debug, U: Debug, Y: Debug>() {}
```

Will `Debug` be added 3 times?

File: src/librustdoc/html/render/mod.rs [hunk 5]
```diff
@@ -96,6 +96,22 @@ crate struct IndexItem {
 #[derive(Debug)]
 crate struct RenderType {
     name: Option<String>,
+    generics: Option<Vec<Generic>>,
+}
+
+/// A type used for the search index.
+#[derive(Debug)]
+crate struct Generic {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:104]: At this point, I think we can simply drop this type and simply keep the string into `RenderType::generics`.

File: src/librustdoc/html/render/cache.rs (hunk 4 above)
Comment (notriddle) [reply src/librustdoc/html/render/cache.rs:263]: Yeah, it does. And the JavaScript search tool actually does take advantage of this. You can see in the search code how it's counting the number of times a generic appears in the search query and in the type-being-checked:

https://github.com/rust-lang/rust/blob/4169b402bb6e5876d7dc5f8cf8b31def451badff/src/librustdoc/html/static/search.js#L361-L383

If you mention Debug twice in your query, then it must appear twice in the type signature.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:262]: Oh I see. Please add a comment about this behaviour then so the next ones reading this code won't ask the same question.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:262]: Please also add a test to check this behaviour (regressions did a lot of bad on the search). I think after that it'll be good. Just need to run the perf check and we're good to go. Thanks a lot for fixing this issue!

Comment (notriddle) [reply src/librustdoc/html/render/cache.rs:262]: Okay, I added a test case for "bag semantics" search results. It also required a bug fix for the query parsing, since it was breaking up `Foo<A, B>, Bar` into three queries, when it needed to be two.

File: src/librustdoc/html/render/cache.rs [hunk 6]
```diff
@@ -251,6 +252,23 @@ fn get_index_type_name(clean_type: &clean::Type, accept_generic: bool) -> Option
     }
 }
 
+// Return a list of generic parameters for use in the search index.
+//
+// This function replaces bounds with types, so that `T where T: Debug` just becomes `Debug`.
+// It does return duplicates, and that's intentional, since search queries like `Result<usize, usize>`
+// are supposed to match only results where both parameters are `usize`.
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/cache.rs:259]: Please turn this comment into a doc comment (very useful when documenting private APIs).

File: src/librustdoc/html/static/search.js [hunk 7]
```diff
@@ -354,10 +357,13 @@ window.initSearch = function(rawSearchIndex) {
                 if (literalSearch) {
                     if (val.generics && val.generics.length !== 0) {
                         if (obj.length > GENERICS_DATA &&
-                              obj[GENERICS_DATA].length >= val.generics.length) {
+                              obj[GENERICS_DATA].length >= 0) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/search.js:360]: Should this be:

```suggestion
                             obj[GENERICS_DATA].length > 0) {
```

?

**Comment by GuillaumeGomez** [comment]
Can you rebase on `master` now that #86746 was merged please?

**Comment by notriddle** [comment]
@GuillaumeGomez kay, it's rebased

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks! Considering that the number of changes on the rustdoc "backend" side is now very small, I don't think we need to run a perf check anymore. r=me once CI pass.
